### ISSUE TITLE
No healthy upstream with egress gateway

### ISSUE THREAD

**Issue by madjam002** [istio/istio#7077]
**Describe the bug**
When following the guide here (https://istio.io/docs/tasks/traffic-management/egress-gateway/#define-an-egress-gateway-and-direct-http-traffic-through-it), trying to reach `edition.cnn.com` results in a `503` error `No healthy upstream`

**Expected behavior**
Running `curl -D - http://edition.cnn.com` should return a non 503 error.

**Steps to reproduce the bug**

Originally I was using kubectl and the istio custom resource definitions, but using istoctl made no difference.

Run:
```
cat <<EOF | istioctl create -f -
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: istio-egressgateway
spec:
  selector:
    istio: egressgateway
  servers:
  - port:
      number: 80
      name: http
      protocol: HTTP
    hosts:
    - "edition.cnn.com"
EOF
```

and

```
cat <<EOF | istioctl create -f -
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: cnn
spec:
  hosts:
  - edition.cnn.com
  ports:
  - number: 80
    name: http-port
    protocol: HTTP
  - number: 443
    name: https
    protocol: HTTPS
  resolution: DNS
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: direct-through-egress-gateway
spec:
  hosts:
  - edition.cnn.com
  gateways:
  - istio-egressgateway
  - mesh
  http:
  - match:
    - gateways:
      - mesh
      port: 80
    route:
    - destination:
        host: istio-egressgateway.istio-system.svc.cluster.local
        port:
          number: 80
      weight: 100
  - match:
    - gateways:
      - istio-egressgateway
      port: 80
    route:
    - destination:
        host: edition.cnn.com
        port:
          number: 80
      weight: 100
EOF
```

Start the sleep container:
```
kubectl apply -f <(istioctl kube-inject -f samples/sleep/sleep.yaml)
```

Exec to the sleep container in the pod and run:
```
$ curl -D - edition.cnn.com
HTTP/1.1 503 Service Unavailable
content-length: 19
content-type: text/plain
date: Fri, 13 Jul 2018 08:45:43 GMT
server: envoy
x-envoy-upstream-service-time: 3

no healthy upstream
```

**Version**
```
$ istioctl version
Version: 0.8.0
GitRevision: 6f9f420f0c7119ff4fa6a1966a6f6d89b1b4db84
User: root@48d5ddfd72da
Hub: docker.io/istio
GolangVersion: go1.10.1
BuildStatus: Clean

$ kubectl version
Client Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.0", GitCommit:"fc32d2f3698e36b93322a3465f63a14e9f0eaead", GitTreeState:"clean", BuildDate:"2018-03-27T00:14:31Z", GoVersion:"go1.9.4", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.4", GitCommit:"5ca598b4ba5abb89bb773071ce452e33fb66339d", GitTreeState:"clean", BuildDate:"2018-06-06T08:00:59Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
```

**Is Istio Auth enabled or not?**
```
helm upgrade --install system-istio \
  install/kubernetes/helm/istio --namespace istio-system
```

**Environment**
Bare metal, Ubuntu 16.04.4, Calico networking

**Additional info**
The istio pilot logs show a lot of these:

```
2018-07-13T08:48:37.827087Z	info	GetService(edition.cnn.com) error: service does not exist
2018-07-13T08:48:37.827513Z	info	GetService(edition.cnn.com) error: service does not exist 
```

The istio egress gateway logs show the 503:
```
[2018-07-13T08:50:40.839Z] "GET / HTTP/1.1" 503 UH 0 19 0 - "10.233.232.16" "curl/7.35.0" "8835afe3-f67b-9ce7-891a-9d0b4b3474f5" "edition.cnn.com" "-"
```

Not sure if that's helpful

**Comment by kHRISl33t** [comment]
usually it means that it cant match for the host. Are you sure that your destination host is correct?

**Comment by ymesika** [comment]
@madjam002 You are doing it right. Just remove the `mesh` gateways entry and the 1st `match` from the `VirtualService` and you are set. Should work.

**Comment by madjam002** [comment]
_sigh_ This was entirely my fault, it was DNS!

I have a bunch of search domains set up, and one of my zone files for my domains was causing the DNS server to return a `SERVFAIL`. For some reason, when any of the istio pods try and resolve any domain (e.g `edition.cnn.com`), it was falling over when `edition.cnn.com.mybrokensearchdomain` was returning a `SERVFAIL`. Oddly enough, none of my other pods from other applications were affected by this, but either way I've fixed the DNS server and it's all working now!

Out of interest @ymesika, why did you suggest removing the `mesh` egress gateway? Now I've fixed the underlying issue, when I tried removing `mesh`, none of my mixer policies were working.
